README.md
<img src="man/figures/logo.png" align="right" width="96" height="96">

# pophelper

[![ci_badge](https://github.com/royfrancis/pophelper/workflows/build/badge.svg)](https://github.com/royfrancis/pophelper/actions?workflow=build) [![codecov_.io_badge](https://codecov.io/github/royfrancis/pophelper/coverage.svg?branch=master)](https://codecov.io/github/royfrancis/pophelper?branch=master)

`pophelper` is an R package and web app to analyse and visualise population structure. See the [project website](http://royfrancis.github.io/pophelper/) for more details. An interactive version of this package using shiny web framework is available as [pophelperShiny](https://github.com/royfrancis/pophelperShiny).

## Disclaimer

This R package is offered free and without warranty of any kind, either expressed or implied. I will not be held liable to you for any damage arising out of the use, modification or inability to use this program. This R package can be used, redistributed and/or modified freely for non-commercial purposes subject to the original source being properly cited. Licensed under GPL-3. Please make sure you verify all your results.  

## Contact

If you have an comments, suggestions, corrections or ideas on ways to improve or extend this package, feel free to contact me. Submit a report on the [Github issues page](https://github.com/royfrancis/pophelper/issues).  

2020 | Roy M Francis  


NEWS
pophelper 2.3.1
=================
* 28-Dec-2020
* Min req R-3.5.
* Error bars in first plot of Evanno method showed max/min rather than SD. Fixed to show SD.
* Input and output directories are to be specified explicitly by user as per CRAN guidelines
* cat and print statements replaced with message or warning
* Argument 'quiet' removed in all functions
* Fixed dependency issues
* Automatic CI testing added
* Fixed an issue where runs merged using mergeQ() produced a warning about missing attributes.
* In function clumppExport() and distructExport(), argument 'path' changed to 'exportpath' to be consistent with other functions.
* Legend is now sorted numerically
* Fixed minor errors in documentation.

pophelper 2.3.0
=================
* 11-Aug-2019
* Min req R-3.5.0.
* Removed CLUMPP and DISTRUCT executables
* 'useexe' argument in ClumppExport() deprecated.
* Added alignK() function for label switching within and between K
* Argument 'clumppexport' in analyseQ() removed and replaced with 'align'
* New dependency package label.swiching
* Updated function is.qlist()
* Added function as.qlist() to correctly format a qlist object
* Added function joinQ() to combine multiple qlist objects
* Added function sortQ() to sort runs in a qlist by a run attribute
* Added function mergeQ() to merge replicate runs within K
* Added function splitQ() to split a qlist object by an attribute
* 14-Aug-2019
* Added arguments 'legendspacing' and 'legendrow' to plotQ()

pophelper 2.2.9
=================
* 14-Apr-2019
* plotQ(): Order of runs in a joined plot was incorrect when k>10 and grplab in use. Fixed.

pophelper 2.2.8.1
=================
* 16-Mar-2019
* Fixed 'exportpath' issues in plotQ()
* Added 'exportpath' to plotQMultiline() and evannoMethod()
* Added 'exportpath' to tabulateQ()
* Added 'exportpath' to summariseQ()

pophelper 2.2.8
=================
* 15-Mar-2019
* Dropped dependency on package cairo
* summariseQ() accepts single run tabulated input
* evannoMethodStructure(): Argument 'file' incorrectly specified for pdf as 'filename'. Fixed
* plotQMultiline(): Several undocumented arguments added to documentation
* Minor corrections to documentation
* Adjustments to getDim()
* add argument exportpath to all functions
* plotQ(): Added argument 'showgrplab'
* readQ() and readQStructure(): Reads confidence intervals from structure files if available

pophelper 2.2.7
=================
* 23-May-2018
* With joined plots, group labels and sorting, div lines were incorrectly printed causing plots to be misaligned. This has been resolved
* Removed restriction of setting 'ordergrp=T' when sorting by cluster or 'all'

pophelper 2.2.6
=================
* 02-May-2018
* getDim() adjusted better for shared and unshared individual labels
* Argument 'grplabface' added to plotQ() to adjust font face of group labels
* plotQ(): Div lines are printed with joined plots, group labels and sorting
* plotQ(): With joined plots and group labels with multiple label sets, when sorting, argument 'ordergrp' must be set to TRUE. Individuals and group labels were not correctly matched with joined output, group labels and sorting. This has been corrected. Group labels are sorted by all labels started with selgrp
* for-loops in all functions changed to seq_along()
* sortInd() gets a new argument 'ordergrp' which orders group labels when sorting

pophelper 2.2.5.1
=================
* 19-Jan-2018
* Added arguments 'xaxisbreaks' and 'xaxislabels' to evannoMethodStructure()

pophelper 2.2.5
=================
* 11-Jan-2018
* A bug in plotQ() caused strip labels to be incorrect. This has been fixed
* plotQMultiline() subtitle by default prints the K value of the run
* plotQ() and plotQMultiline() prints 'Drawing plot ...' when exporting plot

pophelper 2.2.4
=================
* 07-Jan-2018
* Function readQBaps() added to convert BAPS run files to qlist
* Function readQ() modified to read BAPS run files
* Function checkQ() modified to identify BAPS run files
* Open textConnection() fixed in readQStructure()
* Function readQTess() modified to use textConnection()
* Functions readQStructure(),readQTess(),readQBasic() and readQClumpp() 
  are now exported to the namespace.
* tabulateQ() has 'quiet=T' set as default
* summariseQ has 'quiet=T' set as default
* Changes to plotQ()
  + Argument 'sharedindlab' added to control separate vs common labels in 'join' plots
  + Argument 'indlabwithgrplab' added to control if grp labels are concatenated to ind lab
  + Argument 'showyaxis' added to turn y-axis on/off
  + Argument 'showticks' added to turn x and y axes ticks on/off
  + Argument 'ticksize' added to adjust tick size
  + Argument 'ticklength' added to adjust tick length
  + Argument 'basesize' added to adjust text size of all theme elements
  + Argument 'panelratio' added to adjust plot panel to grplab panel ratio
  + Argument 'legendmargin' added to adjust margin around the legend
  + Argument 'exportplot' added to export plot
  + Argument 'returnplot' added to return plot object
  + Argument 'returndata' added to return processed data
  + Default individual labels are added before sorting, subsetting etc
    therefore, individual labels now reflect change is sorting.
  + Fixed a bug in sorting where sorting by 'label' didn't work
  + Argument 'quiet' added to turn off messages
  + Facets are plotted by facet_wrap rather than facet_grid
  + Minor adjustments to margins
  + Div lines are not drawn with 'join' plot, when sorting by all or cluster. (Bug to be fixed)
* Changes to plotQMultiline()
  + Argument 'indlabwithgrplab' added to control if grp labels are concatenated to ind lab
  + Arguments 'showlegend','legendlab','legendpos','legendkeysize','legendtextsize','legendmargin' added
  + Argument 'ticksize' added to adjust tick size
  + Argument 'ticklength' added to adjust tick length
  + Argument 'basesize' added to adjust text size of all theme elements
  + Argument 'quiet' added to turn off messages

tests/testthat.R
library(testthat)
library(pophelper)

test_check("pophelper")


tests/testthat/test-all.R
library(testthat)
library(pophelper)

#devtools::test()

#Preparation
deleteoutput = TRUE
#create a new folder and set as wd
#currwd <- getwd()
#dir.create(paste(currwd,"/PophelperDemo",sep=""))
#setwd(paste(currwd,"/PophelperDemo",sep=""))
#read sample STRUCTURE files from R package
sfiles <- list.files(path=system.file("/files/structure",package="pophelper"),full.names=TRUE)
#read sample TESS files from R package
tfiles <- list.files(path=system.file("/files/tess",package="pophelper"),full.names=TRUE)
#read sample ADMIXTURE files from R package
afiles <- list.files(path=system.file("/files/admixture",package="pophelper"),full.names=TRUE)

#-------------------------------------------------------------------------------

#tabulateRuns
context("Tabulate")
tr1 <- tabulateRunsStructure(files=sfiles)
tr2 <- tabulateRunsTess(files=tfiles)
tr3 <- tabulateRunsAdmixture(files=afiles)

test_that("Is output dataframe?",{
  expect_equal(class(tr1),"data.frame")
  expect_equal(class(tr2),"data.frame")
  expect_equal(class(tr3),"data.frame")
})

test_that("quiet=FALSE",{
  expect_output(tabulateRunsStructure(files=sfiles, quiet=FALSE),"Number of files selected: 17")
  expect_output(tabulateRunsTess(files=tfiles, quiet=FALSE),"Number of files selected: 21")
  expect_output(tabulateRunsAdmixture(files=afiles, quiet=FALSE),"Number of files selected: 10")
})

test_that("quiet=TRUE",{
  expect_output(tabulateRunsStructure(files=sfiles, quiet=TRUE),"")
  expect_output(tabulateRunsTess(files=tfiles, quiet=TRUE),"")
  expect_output(tabulateRunsAdmixture(files=afiles, quiet=TRUE),"")
})

test_that("sorttable=FALSE",{
  expect_equal(row.names(tabulateRunsStructure(files=sfiles, quiet=FALSE,sorttable = FALSE)),
               c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17"))
  expect_equal(row.names(tabulateRunsTess(files=tfiles, quiet=FALSE,sorttable = FALSE)),
               c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"))
  expect_equal(row.names(tabulateRunsAdmixture(files=afiles, quiet=FALSE,sorttable = FALSE)),
               c("1","2","3","4","5","6","7","8","9","10"))
})

# test_that("sorttable=TRUE",{
#   expect_equal(row.names(tabulateRunsStructure(files=sfiles,sorttable = TRUE)), c("1","10","17","2","11","12","3","7","13","4","8","14","5","9","15","6","16"))
#   expect_equal(row.names(tabulateRunsTess(files=tfiles,sorttable = TRUE)), c("1","8","15","2","9","16","3","10","17","4","11","18","5","12","19","6","13","20","7","14","21"))
#   expect_equal(row.names(tabulateRunsAdmixture(files=afiles,sorttable = TRUE)), c("1","2","3","4","5","6","7","8","9","10"))
# })

tabulateRunsStructure(files=sfiles,writetable=TRUE)
tabulateRunsTess(files=tfiles,writetable=TRUE)
tabulateRunsAdmixture(files=afiles,writetable=TRUE)

test_that("writetable=TRUE",{
  expect_equal("tabulateRunsStructure.txt" %in% list.files(),TRUE)
  expect_equal("tabulateRunsTess.txt" %in% list.files(),TRUE)
  expect_equal("tabulateRunsAdmixture.txt" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("tabulateRunsStructure.txt")
if(deleteoutput) file.remove("tabulateRunsTess.txt")
if(deleteoutput) file.remove("tabulateRunsAdmixture.txt")

test_that("Error: no input",{
  expect_error(tabulateRunsStructure())
  expect_error(tabulateRunsTess())
  expect_error(tabulateRunsAdmixture())
})

#-------------------------------------------------------------------------------

#summariseRuns
context("Summarise")
test_that("Is output dataframe?",{
  expect_equal(class(summariseRunsStructure(tr1)),"data.frame")
  expect_equal(class(summariseRunsTess(tr2)),"data.frame")
  expect_equal(class(summariseRunsAdmixture(tr3)),"data.frame")
})

sr1 <- summariseRunsStructure(tr1,writetable=TRUE)
sr2 <- summariseRunsTess(tr2,writetable=TRUE)
sr3 <- summariseRunsAdmixture(tr3,writetable=TRUE)

test_that("writetable=TRUE",{
  expect_equal("summariseRunsStructure.txt" %in% list.files(),TRUE)
  expect_equal("summariseRunsTess.txt" %in% list.files(),TRUE)
  expect_equal("summariseRunsAdmixture.txt" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("summariseRunsStructure.txt")
if(deleteoutput) file.remove("summariseRunsTess.txt")
if(deleteoutput) file.remove("summariseRunsAdmixture.txt")

test_that("Error: no input",{
  expect_error(summariseRunsStructure())
  expect_error(summariseRunsTess())
  expect_error(summariseRunsAdmixture())
})

#-------------------------------------------------------------------------------

#evannoMethodStucture
context("Evanno method")
test_that("Is output dataframe?",{
  expect_equal(class(evannoMethodStructure(sr1)),"data.frame")
})
evannoMethodStructure(sr1,writetable=TRUE)
test_that("writetable=TRUE",{
  expect_equal("evannoMethodStructure.txt" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("evannoMethodStructure.txt")

evannoMethodStructure(sr1,exportplot=TRUE)
test_that("exportplot=TRUE",{
  expect_equal("evannoMethodStructure.png" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("evannoMethodStructure.png")

evannoMethodStructure(sr1,exportplot=TRUE,imgtype="jpeg")
test_that("exportplot=TRUE jpeg",{
  expect_equal("evannoMethodStructure.jpg" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("evannoMethodStructure.jpg")

evannoMethodStructure(sr1,exportplot=TRUE,imgtype="pdf")
test_that("exportplot=TRUE pdf",{
  expect_equal("evannoMethodStructure.pdf" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("evannoMethodStructure.pdf")

evannoMethodStructure(sr1,exportplot=TRUE,height=4,width=4, res=400)
test_that("exportplot=TRUE change dim res",{
  expect_equal("evannoMethodStructure.png" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("evannoMethodStructure.png")

test_that("Error: Two input K",{
  expect_error(evannoMethodStructure(sr1[1:2,]))
})

test_that("Error: Non sequential K",{
  expect_error(evannoMethodStructure(sr1[c(1,2,4),],exportplot=TRUE))
})

test_that("exportplot=TRUE with error",{
  expect_equal("kPlot.png" %in% list.files(),TRUE)
})
if(deleteoutput) file.remove("kPlot.png")

test_that("Warning: runs < 2",{
  expect_warning(evannoMethodStructure(summariseRunsStructure(tr1[1:16,])[4:6,]))
})

test_that("Error: no input",{
  expect_error(evannoMethodStructure())
})

#-------------------------------------------------------------------------------

#runsToDf
context("runsToDf")
test_that("Is output dataframe or list?",{
  expect_equal(class(runsToDfStructure(sfiles)),"list")
  expect_equal(class(runsToDfStructure(sfiles[1])),"data.frame")
  expect_equal(class(runsToDfTess(tfiles)),"list")
  expect_equal(class(runsToDfTess(tfiles[1])),"data.frame")
  expect_equal(class(runsToDfAdmixture(afiles)),"list")
  expect_equal(class(runsToDfAdmixture(afiles[1])),"data.frame")
})

test_that("Error: no input",{
  expect_error(runsToDfStructure())
  expect_error(runsToDfTess())
  expect_error(runsToDfAdmixture())
})
#-------------------------------------------------------------------------------

#clumppExport
context("Clumpp Output")
clumppExportStructure(sfiles)
test_that("clumpp export check",{
  expect_equal(all(grepl("STRUCTUREpop",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("STRUCTUREpop*", recursive = TRUE, force = TRUE)

clumppExportStructure(sfiles,prefix="Boom")
test_that("clumpp export check prefix",{
  expect_equal(all(grepl("Boom",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("Boom*", recursive = TRUE, force = TRUE)

clumppExportTess(tfiles)
test_that("clumpp export check",{
  expect_equal(all(grepl("TESSpop",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("TESSpop*", recursive = TRUE, force = TRUE)

clumppExportTess(tfiles,prefix="Hahaha")
test_that("clumpp export check prefix",{
  expect_equal(all(grepl("Hahaha",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("Hahaha*", recursive = TRUE, force = TRUE)

clumppExportAdmixture(afiles)
test_that("clumpp export check",{
  expect_equal(all(grepl("ADMIXTUREpop",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("ADMIXTUREpop*", recursive = TRUE, force = TRUE)

clumppExportStructure(sfiles,prefix="Nanana")
test_that("clumpp export check prefix",{
  expect_equal(all(grepl("Nanana",list.dirs()[-1])),TRUE)
})
if(deleteoutput) unlink("Nanana*", recursive = TRUE, force = TRUE)

#-------------------------------------------------------------------------------

#collectClumppOutput
context("Collect clumpp output")
# clumppExportStructure(sfiles)
# collectClumppOutput()
# clumppExportStructure(tfiles)
# collectClumppOutput(prefix="TESSpop")
# clumppExportStructure(afiles)
# collectClumppOutput(prefix="ADMIXTUREpop")

#-------------------------------------------------------------------------------

#plotRuns
context("plotRuns Structure")
plotRuns(sfiles[1])
test_that("check output",{
  expect_equal(paste0(basename(sfiles[1]),".png") %in% list.files(),TRUE)
})
if(deleteoutput) file.remove(paste0(basename(sfiles[1]),".png"))

plotRuns(sfiles)
test_that("check output",{
  expect_equal(length(grep("structure\\w+.png$",list.files())),17)
})
if(deleteoutput) file.remove(paste0(basename(sfiles),".png"))

plotRuns(sfiles[1:2],imgoutput="join")
test_that("check output joined",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

pops <- read.delim(system.file("files/structurepoplabels.txt",package="pophelper"),header=FALSE)
test_that("check if pops df",{
  expect_equal(class(pops),"data.frame")
})

plotRuns(sfiles[1],poplab=pops$V1)
test_that("check output sep with labels",{
  expect_equal(paste0(basename(sfiles[1]),".png") %in% list.files(),TRUE)
})
if(deleteoutput) file.remove(paste0(basename(sfiles[1]),".png"))

plotRuns(sfiles[1:2],imgoutput="join",poplab=pops$V1)
test_that("check output joined with labels",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

context("plotRuns Tess")
plotRuns(tfiles[1])
test_that("check output",{
  expect_equal(any(grepl("TESS",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("TESS",list.files())])

plotRuns(tfiles[1:2],imgoutput="join")
test_that("check output joined",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

labs1 <- factor(c(rep("PopA",30),rep("PopB",45)))

plotRuns(tfiles[1],poplab=labs1)
test_that("check output sep with labels",{
  expect_equal(any(grepl("TESS",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("TESS",list.files())])

plotRuns(tfiles[1:2],imgoutput="join",poplab=labs1)
test_that("check output joined with labels",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

context("plotRuns Admixture")
plotRuns(afiles[1])
test_that("check output",{
  expect_equal(any(grepl("adm",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("adm",list.files())])

plotRuns(afiles[1:2],imgoutput="join")
test_that("check output joined",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

admpops<-read.delim(system.file("files/admixturepoplabels.txt",package="pophelper"),header=FALSE)

plotRuns(afiles[1],poplab=admpops$V1)
test_that("check output sep with labels",{
  expect_equal(any(grepl("adm",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("adm",list.files())])

plotRuns(afiles[1:2],imgoutput="join",poplab=admpops$V1)
test_that("check output joined with labels",{
  expect_equal(length(grep("Joined.+",list.files())),1)
})
if(deleteoutput) file.remove(list.files()[grep("Joined*",list.files())])

context("plotRuns Table")

tabs1 <- c(system.file("files/STRUCTUREpop_K4-combined.txt",package="pophelper"),
           system.file("files/STRUCTUREpop_K4-combined-aligned.txt",package="pophelper"),
           system.file("files/STRUCTUREpop_K4-combined-merged.txt",package="pophelper"))

plotRuns(files=tabs1,imgoutput="tab")

test_that("check output table",{
  expect_equal(length(grep("STRUCTUREpop",list.files())),3)
})
if(deleteoutput) file.remove(list.files()[grep("STRUCTUREpop",list.files())])
#-------------------------------------------------------------------------------

#plotMultiline
context("plotMultiline")
plotMultiline(sfiles[1])
test_that("check output",{
  expect_equal(any(grepl("structure",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("structure",list.files())])

plotMultiline(tfiles[1])
test_that("check output",{
  expect_equal(any(grepl("TESS",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("TESS",list.files())])

plotMultiline(afiles[1])
test_that("check output",{
  expect_equal(any(grepl("adm",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("adm",list.files())])

#-------------------------------------------------------------------------------

context("Collect Tess runs")

#-------------------------------------------------------------------------------

context("plotRunsInterpolate")
cfile239 <- read.delim(system.file("files/coords239.txt",package="pophelper"),header=FALSE)
sfile239 <- system.file("files/Structure239_4",package="pophelper")

plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239)
test_that("check output krig",{
  expect_equal(any(grepl("Interpolation-krig",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-krig",list.files())])

plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239,method = "bilinear")
test_that("check output bilinear",{
  expect_equal(any(grepl("Interpolation-bilinear",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-bilinear",list.files())])

plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239,method = "bicubic")
test_that("check output bicubic",{
  expect_equal(any(grepl("Interpolation-bicubic",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-bicubic",list.files())])

plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239,method = "idw")
test_that("check output idw",{
  expect_equal(any(grepl("Interpolation-idw",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-idw",list.files())])

plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239,method = "nn")
test_that("check output nn",{
  expect_equal(any(grepl("Interpolation-nn",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-nn",list.files())])

cd2 <- system.file("/files/coords75.txt",package="pophelper")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2)
test_that("check output krig",{
  expect_equal(any(grepl("Interpolation-krig",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Interpolation-krig",list.files())])

plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,imgtype="jpeg")
test_that("check output krig jpeg",{
  expect_equal(any(grepl("Clusters.jpg",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Clusters.jpg",list.files())])

plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,imgtype="pdf")
test_that("check output krig pdf",{
  expect_equal(any(grepl("Clusters.pdf",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Clusters.pdf",list.files())])

#-------------------------------------------------------------------------------

context("plotRunsSpatial")
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239)
test_that("check output",{
  expect_equal(any(grepl("Spatial.png",list.files())),TRUE)
})
if(deleteoutput) file.remove(list.files()[grep("Spatial.png",list.files())])



inst/files/PophelperDemo.R
#POPHELPER 1.1.6
#02-JAN-2016
#DEMO SCRIPT

library(pophelper)
#create a new folder and set as wd
currwd <- getwd()
dir.create(paste(currwd,"/PophelperDemo",sep=""))
setwd(paste(currwd,"/PophelperDemo",sep=""))
#read sample STRUCTURE files from R package
sfiles <- list.files(path=system.file("/files/structure",package="pophelper"),full.names=T)
#read sample TESS files from R package
tfiles <- list.files(path=system.file("/files/tess",package="pophelper"),full.names=T)
#read sample ADMIXTURE files from R package
afiles <- list.files(path=system.file("/files/admixture",package="pophelper"),full.names=T)
#tabulate STRUCTURE runs
sf1 <- tabulateRunsStructure(files=sfiles, quiet=F)
sf1 <- tabulateRunsStructure(files=sfiles, writetable = T)
sf1 <- tabulateRunsStructure(files=sfiles, writetable = T, sorttable = F, quiet = T)
sf1
#tabulate TESS runs
tf1 <- tabulateRunsTess(files=tfiles, writetable = T)
tf1 <- tabulateRunsTess(files=tfiles, writetable = T, sorttable = F, quiet = T)
tf1
#tabulate ADMIXTURE runs
af1 <- tabulateRunsAdmixture(files=afiles, writetable = T)
af1 <- tabulateRunsAdmixture(files=afiles, writetable = T, sorttable = F, quiet = T)
af1
#summarise STRUCTURE runs
sf2 <- summariseRunsStructure(sf1)
sf2 <- summariseRunsStructure(sf1, writetable = T)
sf2
#summarise TESS runs
tf2 <- summariseRunsTess(tf1, writetable = F)
tf2 <- summariseRunsTess(tf1, writetable = T)
tf2
#summarise ADMIXTURE runs
af2 <- summariseRunsAdmixture(af1, writetable = F)
af2 <- summariseRunsAdmixture(af1, writetable = T)
af2
#perform evanno method
evannoMethodStructure(sf2)
evannoMethodStructure(sf2, writetable = T, exportplot = T)
evannoMethodStructure(sf2, writetable = T, exportplot = T, height = 15, width = 15)
evannoMethodStructure(sf2, height = 20, width = 20, basesize=8)
#convert STRUCTURE run files to R dataframe
runsToDfStructure(files=sfiles)
#convert TESS run files to R dataframe
runsToDfTess(files=tfiles)
#convert ADMIXTURE run files to R dataframe
runsToDfAdmixture(files=afiles)
#convert STRUCTURE runs for CLUMPP
clumppExportStructure(files=sfiles)
clumppExportStructure(files=sfiles, prefix = "Boom", parammode = 3, paramrep = 200)
#convert TESS runs for CLUMPP
clumppExportTess(files=tfiles)
#convert ADMIXTURE runs for CLUMPP
clumppExportAdmixture(files=afiles)
#collect CLUMPP outputs into single directory
#collectClumppOutput(prefix="STRUCTUREpop", filetype="both")  
#collectClumppOutput(prefix="TESSpop", filetype="both")
#CANNOT TEST ABOVE WITHOUT CLUMPP.EXE
#plot separate figures STRUCTURE
plotRuns(files=sfiles[1:2])
plotRuns(files=sfiles[1:2], imgoutput = "sep", height = 5, width = 15) 
#plot separate figures TESS
plotRuns(files=tfiles[1:2]) 
#plot joined figures STRUCTURE
plotRuns(files=sfiles[1:2], imgoutput="join",na.rm=T)
#plot joined figures TESS
plotRuns(files=tfiles[1:2], imgoutput="join")
#read labels for STRUCTURE
pops<-read.delim(system.file("files/structurepoplabels.txt",package="pophelper"),header=F)
#plot separately with labels
plotRuns(files=sfiles[1:2], imgoutput="sep", poplab=pops$V1) 
#plot joined with labels
plotRuns(files=sfiles[1:2], imgoutput="join", poplab=pops$V1) 
#create TESS labels
labs1 <- factor(c(rep("PopA",30),rep("PopB",45)))
#plot TESS runs with labels
plotRuns(files=tfiles[1:2], imgoutput="sep", poplab=labs1)
plotRuns(files=tfiles[1:2], imgoutput="join", poplab=labs1)
#plot ADMIXTURE runs 
plotRuns(files=afiles[1:2], imgoutput="sep")
plotRuns(files=afiles[1:2], imgoutput="join")
#read labels for ADMIXTURE
admpops<-read.delim(system.file("files/admixturepoplabels.txt",package="pophelper"),header=F)
plotRuns(files=afiles[1:2], imgoutput="sep", poplab=admpops$V1)
plotRuns(files=afiles[1:2], imgoutput="join", poplab=admpops$V1)
#plot TABLE files
tabs1 <- c(system.file("files/STRUCTUREpop_K4-combined.txt",package="pophelper"),
           system.file("files/STRUCTUREpop_K4-combined-aligned.txt",package="pophelper"),
           system.file("files/STRUCTUREpop_K4-combined-merged.txt",package="pophelper"))
#plot one TABLE file
plotRuns(files=tabs1[1],imgoutput="tab")
#plot all TABLE files
plotRuns(files=tabs1,imgoutput="tab")
#plot all TABLE files with labels
plotRuns(files=tabs1,imgoutput="tab",poplab=pops$V1)
#plot multiline STRUCTURE
plotMultiline(files=sfiles[1])
#plot multiline TESS
plotMultiline(files=tfiles[1])
#plot multiline ADMIXTURE
plotMultiline(files=afiles[1])
#plot multiline ADMIXTURE multiple
plotMultiline(files=afiles[2:3])
#plotmultiline with custom setting
plotMultiline(files=sfiles[1], spl=75, lpp=10)
#plotmultiline with custom setting
plotMultiline(files=sfiles[1],barspace = 0,barwidth = 1,height = 6,width=16)
#plotmultiline with custom setting
plotMultiline(files=sfiles[1],barspace = 0,barwidth = 1,height = 6,width=16,yaxislabs = T,indlabs = F)
#plotmultiline with custom setting
plotMultiline(files=sfiles[1],barspace = 0,barwidth = 1,height = 6,width=16,yaxislabs = T,labangle = 45,imgtype = "jpeg")
#plotMultiline with one TABLE file
plotMultiline(files=tabs1[1])
#plotMultiline with all TABLE file
plotMultiline(files=tabs1)
#read coordinate file
cd2 <- system.file("/files/coords75.txt",package="pophelper")
#basic usage
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2)
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,imgtype="jpeg")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,imgtype="pdf")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,method="bicubic")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,method="krig")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,clusters=2)
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,pointcol = "red")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,clusters=c(1,3))
#adjusting legend size and legend text
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,legendsize=0.4,legendtextsize=6,method="bilinear")
#removing legend and changing method
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,legend=FALSE,method="idw")
plotRunsInterpolate(datafile=tfiles[2],coordsfile=cd2,legend=FALSE,method="idw",idwpower=5)
#read coordinate file length 239
cfile239 <- read.delim(system.file("files/coords239.txt",package="pophelper"),header=F)
#select path to STRUCTURE file length 239
sfile239 <- system.file("files/Structure239_4",package="pophelper")
#basic usage
plotRunsInterpolate(datafile=sfile239,coordsfile=cfile239,cluster=1:2,height=8)
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239)
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,imgtype="jpeg")
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,imgtype="pdf")
#needs more height. set height and width as required in cm.
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12)
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,showaxis = T)
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,showaxis = T, popcol=c("blue","green","red","steelblue","yellow"))
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,showaxis = T, pointcol ="red")
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,pointtype=12)
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,chull=T,chullsize=2,ellipse=F,)
#set UTM coordinates. Better geographic distance representation over a scale such as normal countries.
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,setutm=T)
#without ellipses
plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,height=12,ellipse=F)

#Create a 2x2 montage with varying parameters
#don't export, export data, add title
p1 <- plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,exportplot=F,dataout=T, plottitle="Fig 1")
#without ellipse, with square points and transparency added
p2 <- plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,exportplot=F,dataout=T, plottitle="Fig 2",ellipse=F,pointtype=15,pointtransp=0.4)
#without ellipse, with convex hulls, coordinates in UTM
p3 <- plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,exportplot=F,dataout=T, plottitle="Fig 3",ellipse=F,chull=T, setutm=T)
#no ellipse, with convex hull, decreased convexhull transparency, change cluster labels
p4 <- plotRunsSpatial(datafile=sfile239,coordsfile=cfile239,exportplot=F,dataout=T, plottitle="Fig 4",ellipse=F,chull=T,chulltransp=0.2,chulltype=3,legendlabels=c("PopA","PopB","PopC","PopD","PopE"))

#install gridExtra first
#library(gridExtra)
#png("PlotRunsSpatial.png",height=20,width=20,res=250,units="cm",type="cairo")
#grid.arrange(p1,p2,p3,p4,nrow=2,ncol=2)
#dev.off()
setwd(currwd)
dir.create(paste(currwd,"/AnalyseDemo",sep=""))
setwd(paste(currwd,"/AnalyseDemo",sep=""))

#perform a routine analysis using a wrapper function
analyseRuns(sfiles)
analyseRuns(tfiles)
analyseRuns(afiles)

setwd(currwd)
